### DISCUSSION TITLE
Test rack-protection with rack-test main branch

### DISCUSSION THREAD

**Discussion by jeremyevans** [sinatra/sinatra#1782] [General]
rack-protection currently has a development dependency on rack-test.  We are getting close to releasing a new major version of rack-test, and would like feedback on whether rack-protection will works with the rack-test main branch.  Can the rack-protection maintainer run the tests with the rack-test main branch and let me know whether there are any problems?

Thanks,
Jeremy

**Comment by jkowens** [comment]
Hi @jeremyevans I just now saw your post. I had to lock rack-test < 2 for our CI to pass. That seems to work for now, but it would be nice to dig in and figure out why v2 results in an infinite loop.

**Reply by jeremyevans** [reply]
The cause of the problem is https://github.com/sinatra/sinatra/blob/cf404526373298f06a59e1a129b6cdf8282ae216/sinatra-contrib/lib/sinatra/test_helpers.rb#L212-L214

In rack-test 2, Rack::Test::Session and Rock::MockSession were combined into one class, `rack_mock_session` and `rack_test_session` are aliases, and `rack_test_session` calls `build_rack_test_session`, hence the infinite loop.

You could fix the infinite loop by changing to:

```ruby
    def build_rack_test_session(_name)
      Session.new app
    end
```

That's not the only issue, though.  Another issue you will run into is https://github.com/sinatra/sinatra/blob/cf404526373298f06a59e1a129b6cdf8282ae216/sinatra-contrib/lib/sinatra/test_helpers.rb#L24-L26

The `default_env` private method no longer exists.  You may want to override `env_for`:

```ruby
  def env_for(uri, env)
    env = global_env.merge(env)
    super
  end
```

Hopefully those are the only two issues.  I didn't run the tests to confirm.

**Reply by jkowens** [reply]
Awesome, thank you so much @jeremyevans!
